rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,171 @@
+
+Upon start, the service will write a SSH config to a configured destination.
+This SSH config should be used by OpenSSH to leverage the service.
+
+Once initialized, the service will open a Unix socket. Upon connection to this
+socket, the client will write a short message containing the target host and
+port. The service will then establish the connection, and then begin forwarding
+data between the socket and the upstream connection.
+
+The client will be configured as the `ProxyCommand`, and additionally
+`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
+to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
+actively forwarding the connection. This means the client can exit after the
+connection to the local proxy is established.
+
+This solution means that whilst a process will be spun up for each SSH
+connection, it will return after the connection is established. This reduces the
+long-term overhead associated with each connection and the impact on the Linux
+scheduler.
+
+#### Protocol
+
+The Unix socket will implement a simple protocol:
+
+1. Upon connection, the client will send a message indicating the target.
+  a. The message will be encoded in JSON. This will allow more fields to be 
+    added in the future without breaking compatability.
+  b. The `host_port` field will be a string in the format `host:port`.
+2. The server will establish a connection to the target, and then begin
+   forwarding data between the local connection and the target.
+
+Also considered was implementing the SOCKS5 protocol. Whilst this is more 
+standardized, it is more complicated than necessary. It also relies on our
+existing implementation of a SOCKS5 server which has not received signficant
+attention in some time.
+
+#### Client Binary
+
+As it stands today, the `tbot` binary has significant overhead associated with
+it.
+
+Due to this, we will introduce a new, extremely lightweight binary that will
+act as the client.
+
+The client will accept two arguments:
+
+1. The path to the Unix socket.
+2. The message to send to the `tbot` proxy service. This will have the host and
+   port templated into it by OpenSSH with the `%h` and `%p` placeholders.
+
+Comparing Rust and Go for this lightweight binary:
+
+- Rust has a smaller binary size: 400KiB vs 1.8MiB.
+- Rust executes faster.
+- Tuning Go's GC behaviour requires also using `env` as part of the
+  ProxyCommand. This is not necessary with Rust.
+
+For this binary, we will use Rust.
+
+### Initial Scope
+
+For the initial version, we must include:
+
+- The long-lived proxying background service within `tbot`
+- The new lightweight client binary with support for ProxyUseFdPass.
+- A configuration option to control the maximum number of proxied connections.
+  - When this maximum is reached, the service will wait until the number of
+    active connections drops below the maximum.
```

Comment:
How is this in any way practical, with `ControlMaster`/`ControlPath`? Connections won't get freed until minutes later, and at that point you've basically just rejected the connection.

Reply:
Do you feel that we would be better off removing any form of limitation on the client side ? Or should we be more aggressive and reject the connections rather than holding off ?

Reply:
Hard to say. Rejecting connections would be better, but blocking will let a bunch of short-lived connection through while respecting the concurrency limit, and both options can be good depending on the workload.

Do we need a connection limit?

Reply:
Ended up dropping the limit from implementation - I'll remove from RFD.

Reply:
[9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)